### PR TITLE
AICS service implementation

### PR DESCRIPTION
**Pull request by c-ryan-k**
Adds support for AICS service commands inside a new `az iot product` namespace

---
This project has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/). For more information see the [Code of Conduct FAQ](https://opensource.microsoft.com/codeofconduct/faq/) or contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additional questions or comments.

Thank you for contributing to the IoT extension!

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] If introducing new functionality or modified behavior, are they backed by unit and integration tests?
- [x] In the same context as above are command names and their parameter definitions accurate? Do help docs have sufficient content?
- [x] Have **all** unit **and** integration tests passed locally? i.e. `pytest <project root> -vv`
- [x] Have static checks passed using the .pylintrc and .flake8 rules? Look at the CI scripts for example usage.

### CODE REVIEW

File: azext_iot/product/providers/aics.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+)
+from msrestazure.azure_exceptions import CloudError
+from azext_iot.common.utility import unpack_msrest_error
+from knack.log import get_logger
+from knack.util import CLIError
+
+logger = get_logger(__name__)
+
+
+def process_cloud_error(func):
```
Comment (digimaun) [inline azext_iot/product/providers/aics.py:21]: This is an EXCELLENT idea

File: .gitignore [hunk 2]
```diff
@@ -319,3 +319,6 @@ src/build
 
 # Code coverage
 .coverage
+
+# remove later
+PnPCert
```
Comment (digimaun) [inline .gitignore]: When is later?

File: azext_iot/product/_help.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    """
+    helps[
+        "iot product init"
+    ] = """
+        type: command
+        short-summary: Used to initialize local workspace for a new product certification
+        examples:
+        - name: Basic usage
+          text: >
+            az iot product init --product-name {product_name}
```
Comment (digimaun) [inline azext_iot/product/_help.py:45]: I would explicitly add something to the basic usage example about the default working folder if one isn't specified

File: azext_iot/product/_help.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        type: group
+        short-summary: Manage product certification requirements
+    """
+    helps[
+        "iot product requirement list"
+    ] = """
+        type: command
+        short-summary: Get certification requirements
+        long-summary: |
+                      Discover information about provisioning attestation methods that are supported for each badge type
```
Comment (digimaun) [inline azext_iot/product/_help.py]: Just a note (not necessarily suggesting a change) currently long summary does not get rendered on the MS docs reference page - only short summary.

File: .gitignore (hunk 2 above)
Comment (montgomp) [reply .gitignore]: "When will then be now?" "Soon." I have a PR on my fork for this.

File: azext_iot/product/_help.py (hunk 3 above)
Comment (digimaun) [reply azext_iot/product/_help.py:45]: by add "something" i meant to the example description

File: azext_iot/product/command_product.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+logger = get_logger(__name__)
+
+
+def initialize_workspace(cmd, product_name, working_folder="PnPCert"):
+    id = uuid4()
+    if not os.path.exists(working_folder):
+        os.mkdir(working_folder)
+
+    product_config = {
```
Comment (digimaun) [inline azext_iot/product/command_product.py:18]: Not a blocker but in the future I think we should move this product config somewhere out of this module.

File: azext_iot/product/params.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+from azext_iot.product.shared import BadgeType
+
+
+def load_product_params(self, _):
+    with self.argument_context("iot product") as c:
+        c.argument(
+            "working_folder",
+            options_list=["--working-folder", "--wf"],
+            help="The folder to create in current path",
+            arg_group="IoT Device Certification",
```
Comment (digimaun) [inline azext_iot/product/params.py:21]: yay arg groups

File: azext_iot/product/providers/__init__.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+__all__ = ["aics_service_factory", "AICSServiceProvider"]
+
+
+def aics_service_factory(cmd, base_url):
+    creds = AICSAuthentication(cmd=cmd, base_url=base_url or BASE_URL)
+    api = AICSAPI(base_url=base_url or BASE_URL, credentials=creds)
+
+    api.config.add_user_agent(USER_AGENT)
+    api.config.add_user_agent("azcli/aics/{}".format(cli_version))
```
Comment (digimaun) [inline azext_iot/product/providers/__init__.py:20]: What do we get out of the second user agent that isn't already in the first?

File: azext_iot/product/test/params.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+            "wait",
+            options_list=["--wait", "-w"],
+            help='Wait until test run until status is "started" or "running"',
+            arg_group="IoT Device Certification",
+            arg_type=get_three_state_flag(),
+        )
+        c.argument(
+            "poll_interval",
+            options_list=["--poll-interval", "--interval"],
+            help="Used in conjunction with --wait. Sepcifies how frequently (in seconds) polling occurs",
```
Comment (digimaun) [inline azext_iot/product/test/params.py]: Misspelling of 'Specifies'

File: azext_iot/product/test/params.py [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+        c.argument(
+            "run_id",
+            options_list=["--run-id", "-r"],
+            help="The generated Id of a test run",
+            arg_group="IoT Device Certification",
+        )
+        c.argument(
+            "wait",
+            options_list=["--wait", "-w"],
+            help='Wait until test run until status is "started" or "running"',
```
Comment (digimaun) [inline azext_iot/product/test/params.py]: Maybe too much 'until' in this sentence. How about 'Wait on the test run until the status is "started" or "running"'?

File: azext_iot/product/providers/__init__.py (hunk 7 above)
Comment (montgomp) [reply azext_iot/product/providers/__init__.py:20]: Was a request from Service team to include that format

File: azext_iot/product/providers/auth.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+        session = session or super(AICSAuthentication, self).signed_session()
+        session.headers["Authorization"] = self.generate_token()
+        return session
+
+    def generate_token(self):
+        from azure.cli.core._profile import Profile
+
+        profile = Profile(cli_ctx=self.cmd.cli_ctx)
+        creds, subscription, tenant = profile.get_raw_token()
```
Comment (digimaun) [inline azext_iot/product/providers/auth.py:49]: No resource Id (audience claim on the receiving end) is needed?

Comment (montgomp) [reply azext_iot/product/providers/auth.py:49]: We haven't needed it on any of the AICS environments.

Comment (digimaun) [reply azext_iot/product/providers/auth.py:49]: This is weird because Matt was saying they accept certain audiences in the tokens.

Comment (digimaun) [reply azext_iot/product/providers/auth.py:49]: Resolved - if no resource Id is explicitly added, then https://management.core.windows.net/ is used which Matt said is an allowed audience.

**Pull request merged**